File: runtimes/swift-5.5/Sources/Runtime/RuntimeLogger.swift
import Foundation

class RuntimeLogger {
    static let TYPE_LOG = "log"
    static let TYPE_ERROR = "error"

    var enabled = false
    var id = ""

    var streamLogs: FileHandle? = nil
    var streamErrors: FileHandle? = nil

    init(status: String, id: String) {
        if(status == "enabled" || status == "") {
            self.enabled = true
        }

        if(self.enabled) {
            if(id == "") {
                let serverEnv = ProcessInfo.processInfo.environment["OPEN_RUNTIMES_ENV"]
                if(serverEnv == "development") {
                    self.id = "dev"
                } else {
                    self.id = self.generateId()
                }
            } else {
                self.id = id
            }

            let logsUrl = URL(fileURLWithPath:"/mnt/logs/" + self.id + "_logs.log")
            let errorsUrl = URL(fileURLWithPath:"/mnt/logs/" + self.id + "_errors.log")

            if !FileManager.default.fileExists(atPath: logsUrl.path) {
                FileManager.default.createFile(atPath: logsUrl.path, contents: nil, attributes: nil)
            }

            if !FileManager.default.fileExists(atPath: errorsUrl.path) {
                FileManager.default.createFile(atPath: errorsUrl.path, contents: nil, attributes: nil)
            }

            self.streamLogs = try! FileHandle(forWritingTo: logsUrl)
            self.streamErrors = try! FileHandle(forWritingTo: errorsUrl)

            if let stream = self.streamLogs {
                stream.seekToEndOfFile()
            }

            if let stream = self.streamErrors {
                stream.seekToEndOfFile()
            }
        }
    }

    func write(message: Any, type: String = RuntimeLogger.TYPE_LOG) {
        if(self.enabled == false) {
            return
        }

        var stream = self.streamLogs;

        if(type == RuntimeLogger.TYPE_ERROR) {
            stream = self.streamErrors;
        }

        var stringLog = ""
        if message is CollectionType {
            if let data = try? JSONSerialization.data(withJSONObject: message),
                let stringTemp = String(data: data, encoding: .utf8) {
                    stringLog = stringTemp
            }
        }
        
        if(stringLog == "") {
            stringLog = String(describing: message)
        }

        stringLog += "\n"

        if let stringLogTemp = stringLog.data(using: .utf8) {
            if let streamTemp = stream {
                streamTemp.write(stringLogTemp)
            }
        }
    }

    func end() {
        if(!self.enabled) {
            return;
        }

        self.enabled = false;

        if let stream = self.streamLogs {
            stream.closeFile()
        }

        if let stream = self.streamErrors {
            stream.closeFile()
        }
    }

    func generateId(padding: Int = 7) -> String {
        let now = Date()
        let secs = Int(now.timeIntervalSince1970)
        let usec = Int((now.timeIntervalSince1970 - Double(secs)) * 1_000_000)
        let baseId = String(format: "%08x%05x", secs, usec)
        let randomPadding = (1...padding).map {
            _ in String(format: "%x", Int.random(in: 0..<16))
        }.joined()
        return baseId + randomPadding
    }
}

File: runtimes/dotnet/versions/latest/src/RuntimeLogger.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Text.Json;

namespace DotNetRuntime
{
    public class RuntimeLogger
    {
        public static String TYPE_LOG = "log";
        public static String TYPE_ERROR = "error";

        public String id = "";
        private bool enabled = false;
        private bool includesNativeInfo = false;

        private StreamWriter? streamLogs = null;
        private StreamWriter? streamErrors = null;

        private TextWriter? nativeLogsCache = null;
        private TextWriter? nativeErrorsCache = null;

        private StringBuilder? customStdStream = null;

        public RuntimeLogger(String? status, String? id)
        {
            if (string.IsNullOrEmpty(status) || status == "enabled")
            {
                this.enabled = true;
            }
            else
            {
                this.enabled = false;
            }

            if (this.enabled)
            {
                if (string.IsNullOrEmpty(id))
                {
                    String? serverEnv = Environment.GetEnvironmentVariable("OPEN_RUNTIMES_ENV");
                    if (!string.IsNullOrEmpty(serverEnv) && serverEnv == "development")
                    {
                        this.id = "dev";
                    }
                    else
                    {
                        this.id = this.GenerateId();
                    }
                }
                else
                {
                    this.id = id;
                }

                // Log stream
                this.streamLogs = new StreamWriter("/mnt/logs/" + this.id + "_logs.log", true);
                this.streamErrors = new StreamWriter("/mnt/logs/" + this.id + "_errors.log", true);
            }
        }

        public void Write(object message, String? type = null, bool native = false)
        {
            if (string.IsNullOrEmpty(type))
            {
                type = RuntimeLogger.TYPE_LOG;
            }

            if (this.enabled == false)
            {
                return;
            }

            if (native && !this.includesNativeInfo)
            {
                this.includesNativeInfo = true;
                this.Write(
                    "Native logs detected. Use context.Log() or context.Error() for better experience.",
                    type,
                    native
                );
            }

            var stream = this.streamLogs;

            if (type == RuntimeLogger.TYPE_ERROR)
            {
                stream = this.streamErrors;
            }

            String stringLog = "";
            if (message is IList || message is IDictionary)
            {
                stringLog = JsonSerializer.Serialize(message);
            }
            else
            {
                stringLog = message.ToString() ?? "";
            }

            if (stream != null)
            {
                stream.WriteLine(stringLog);
            }
        }

        public void End()
        {
            if (!this.enabled)
            {
                return;
            }

            this.enabled = true;

            if (this.streamLogs != null)
            {
                this.streamLogs.Close();
            }
            if (this.streamErrors != null)
            {
                this.streamErrors.Close();
            }
        }

        public void OverrideNativeLogs()
        {
            this.nativeLogsCache = Console.Out;
            this.nativeErrorsCache = Console.Error;
            this.customStdStream = new StringBuilder();
            var customStdWriter = new StringWriter(this.customStdStream);
            Console.SetOut(customStdWriter);
            Console.SetError(customStdWriter);
        }

        public void RevertNativeLogs()
        {
            if (this.nativeLogsCache != null)
            {
                Console.SetOut(this.nativeLogsCache);
            }

            if (this.nativeErrorsCache != null)
            {
                Console.SetError(this.nativeErrorsCache);
            }

            if (this.customStdStream != null)
            {
                if (!string.IsNullOrEmpty(this.customStdStream.ToString()))
                {
                    this.Write(this.customStdStream.ToString(), RuntimeLogger.TYPE_LOG, true);
                }
            }
        }

        private String GenerateId(int padding = 7)
        {
            var now = DateTime.UtcNow;
            var epoch = (now - new DateTime(1970, 1, 1));
            var sec = (long)epoch.TotalSeconds;
            var usec = (long)((epoch.TotalMilliseconds * 1000) % 1000);
            var baseId = sec.ToString("x") + usec.ToString("x").PadLeft(5, '0');

            var random = new Random();
            var randomPadding = "";

            for (int i = 0; i < padding; i++)
            {
                var randomHexDigit = random.Next(0, 16).ToString("x");
                randomPadding += randomHexDigit;
            }

            return baseId + randomPadding;
        }
    }
}
